### ISSUE TITLE
Serialization empty List with wrong key 

### ISSUE THREAD

**Issue by xiangning17** [russhwolf/multiplatform-settings#160]
![image](https://github.com/russhwolf/multiplatform-settings/assets/27916852/4717f27b-4a3e-4bc0-9fc3-420e86f383ea)
this line, if meet a empty list, will remove the 0 layer key.

**Comment by russhwolf** [comment]
Do you have an example of something you're trying to do that's not working? This test passes for me so I'm not clear on what the issue is.
```
    @Test
    fun issue_160() {
        val settings: Settings = MapSettings()
        val serializer = ListSerializer(String.serializer())

        settings.encodeValue(serializer, "list", emptyList())

        val list = settings.decodeValueOrNull(serializer, "list")
        assertEquals(emptyList(), list)
    }
```

**Comment by xiangning17** [comment]
First, i define a class:

`@Serializable  data class User(val name: String, val phone: String)`

Then, try this test:
```
    @Test
    fun issue_160() {
        val key = "list"
        val serializer = ListSerializer(User.serializer())
        val (encoder, decoder) = getSettingEncoderAndDecoder(settings, key)

        val origin = emptyList<User>()
        serializer.serialize(encoder, origin)
        var list = serializer.deserialize(decoder)
        assertEquals(origin, list)

        val origin2 = listOf(User("you", "123"), User("me", "234"))
        serializer.serialize(encoder, origin2)
        list = serializer.deserialize(decoder)
        assertEquals(origin2, list)
    }
```
The key to this question is whether the SettingEncoder And SettingsDecoder can be reused. A typical usage scenario is when a delegate is used to declare a property delegate, as shown below:

```
    var users: List<User> by settings.serializedValue(
        ListSerializer(User.serializer()),
        defaultValue = emptyList()
    )
```

**Comment by russhwolf** [comment]
Ok, I still don't understand exactly what you're doing (What is `getSettingEncoderAndDecoder()`? The Multiplatform Settings serialization APIs aren't designed for you to access the encoder and decoder directly) but I can replicate some decoder reuse issues that aren't being triggered by the current tests when using the delegate APIs. The root cause of the issues I can see is that if we throw inside `deserializeOrElse()`, we don't reset state correctly. I'm not 100% sure if this is the same issue you're encountering, but it seems plausible.

I'm surprised this never came up previously. Will fix for the next version.

**Comment by xiangning17** [comment]
Yes, you are right, I just want to explain the problem of Encoder be reused
. `getSettingEncoderAndDecoder ` is not a API, i just use it to help me get the Encoder and Decoder that you declared as private, so as to complete this Test code to illustrate the problems that will arise from Encoder reused. The real problem scenario is when using the Delegate API.

**Comment by russhwolf** [comment]
This should be fixed in 1.1.0
